### PR TITLE
Form tipsets

### LINKED ISSUE #758: Form a tipset with other blocks received at the same height
We receive blocks over pubsub. Ultimately, we extend the chain by adding a tipset (1+ blocks at the same height). We need to implement logic to group received blocks together into a tipset. 

Acceptance Criteria 
- A tipset is formed with blocks at the same height. Any blocks received at the same height (before the next epoch) are added to the tipset.

Notes
- Blocks will not all arrive at the same time, so that must be taken into account.
- This _task may not be required_ if informnewblock is modified such that it processes received blocks (at the same height) in parallel and groups them together in a tipset after processing them.

### PR DESCRIPTION
**Pull request by timvermeulen**
**Summary of changes**
Changes introduced in this pull request:
- Add the `TipsetTracker` type that tracks blocks and can form tipsets from them
- Minor cleanup

**Reference issue to close (if applicable)**

Closes #758 

**Other information and links**

### CODE REVIEW

File: blockchain/chain/src/store/chain_store.rs [hunk 1]
```diff
@@ -102,6 +102,9 @@ pub struct ChainStore<DB> {
 
     /// Used as a cache for tipset lookbacks.
     chain_index: ChainIndex<DB>,
+
+    /// Tracks blocks for the purpose of forming tipsets.
+    tipset_tracker: RwLock<TipsetTracker<DB>>,
```
Comment (austinabell) [inline blockchain/chain/src/store/chain_store.rs:107]: I think it would be best to only wrap the type that interior mutability with the mutex, instead of wrapping the whole tracker type. Also, what are your thoughts on using something like https://github.com/xacrimon/dashmap for a threadsafe map, instead of using mutexes here?

File: blockchain/chain/src/store/tipset_tracker.rs [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+                    log::warn!(
+                        "Have multiple blocks from miner {} at height {} in our tipset cache {}-{}",
+                        header.miner_address(),
+                        header.epoch(),
+                        header.cid(),
+                        cid
+                    );
+                }
+            }
+        }
```
Comment (austinabell) [inline blockchain/chain/src/store/tipset_tracker.rs:46]: This seems like it might be worth to split up these checks into two loops, I would think it might be nice to short circuit the cids checks, so if any are the same, you don't have to load and deserialize a whole block header from storage. 

The second check also seems like a good candidate for a cache in the future (no need to do now) but may be worth it to cache miner addresses -> block cids, since entries are never removed? What are your thoughts?

File: blockchain/chain/src/store/tipset_tracker.rs [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+        entries.push(header.cid().clone());
+    }
+
+    /// Expands the given block header into the largest possible tipset by
+    /// combining it with known blocks at the same height with the same parents.
+    pub fn expand(&self, header: &BlockHeader) -> Result<Tipset, Error> {
+        let mut headers = vec![header.clone()];
```
Comment (austinabell) [inline blockchain/chain/src/store/tipset_tracker.rs:54]: why not just have the function signature take in an owned header, to potentially avoid a clone if the header can be moved in the future?

File: blockchain/chain_sync/src/sync.rs [hunk 4]
```diff
@@ -137,6 +137,137 @@ where
         self.worker_state.clone()
     }
 
+    async fn handle_network_event(
```
Comment (austinabell) [inline blockchain/chain_sync/src/sync.rs:140]: out of curiosity, why was this moved into a function, just to clean up or was there a functional need?

File: blockchain/chain/src/store/chain_store.rs [hunk 5]
```diff
@@ -385,7 +397,7 @@ where
     }
 
     /// Constructs and returns a full tipset if messages from storage exists - non self version
-    pub fn fill_tipset(&self, ts: Arc<Tipset>) -> Result<FullTipset, Arc<Tipset>>
+    pub fn fill_tipset(&self, ts: &Tipset) -> Option<FullTipset>
```
Comment (austinabell) [inline blockchain/chain/src/store/chain_store.rs:400]: ah oops good call this is no longer needed, was an artifact from when Tipset was passed in and was not in wrapped in an Arc and this was a benefit

File: blockchain/chain_sync/src/sync.rs (hunk 4 above)
Comment (timvermeulen) [reply blockchain/chain_sync/src/sync.rs:140]: The direct reason I did this was that rust-analyzer isn't functioning properly inside the `select!` macro. But I think this probably makes the code more readable regardless

File: blockchain/chain/src/store/tipset_tracker.rs (hunk 3 above)
Comment (timvermeulen) [reply blockchain/chain/src/store/tipset_tracker.rs:54]: Oh yeah, I thought we would need to clone here regardless, but I can just extract the epoch first. Good catch

File: blockchain/chain/src/store/tipset_tracker.rs (hunk 2 above)
Comment (timvermeulen) [reply blockchain/chain/src/store/tipset_tracker.rs:46]: Heh, I started out caching miner addresses to avoid having to fetch the entire block header from the blockstore. In the end I reverted to this because I don't know how crucial it is for this tracker to minimize its memory footprint.

The `expand` method could potentially benefit from caching the parent tipset keys as well in order to avoid fetching block headers with different parents, but again, I'm unsure about the memory/speedup trade-offs.

Comment (austinabell) [reply blockchain/chain/src/store/tipset_tracker.rs:46]: Yeah, let's leave it as is for now, but something for us to keep in mind

File: blockchain/chain/src/store/chain_store.rs (hunk 1 above)
Comment (timvermeulen) [reply blockchain/chain/src/store/chain_store.rs:107]: To me this seemed like the more natural place for the lock because it's only in `ChainStore` that the need for concurrent writing arises, `TipsetTracker` in isolation doesn't really care about any of this. e.g. if at some point we hypothetically need to use this tracker without the need for concurrency, locks would be unnecessary. Is that fair?

dashmap is probably faster here, though I don't know if it also optimizes for many concurrent readers like `RwLock` does? I'll leave a note about possibly improving this, at the moment I'm unsure about pulling in a dependency without knowing how much of a speedup it gives us.

File: blockchain/chain/src/store/chain_store.rs [hunk 6]
```diff
@@ -152,14 +156,22 @@ where
...
+    pub async fn add_to_tipset_tracker(&self, header: &BlockHeader) {
+        self.tipset_tracker.write().await.add(header);
+    }
+
     /// Writes tipset block headers to data store and updates heaviest tipset
     pub async fn put_tipset(&self, ts: &Tipset) -> Result<(), Error> {
         persist_objects(self.blockstore(), ts.blocks())?;
-        // TODO determine if expanded tipset is required; see https://github.com/filecoin-project/lotus/blob/testnet/3/chain/store/store.go#L236
-        self.update_heaviest(ts).await?;
+        let expanded = self.expand_tipset(ts.min_ticket_block().clone()).await?;
```
Comment (austinabell) [inline blockchain/chain/src/store/chain_store.rs:166]: this seems pretty redundant, if the heaviest is based on the expanded, why you need to persist the tipset being passed in? Shouldn't this be able to use the other blocks in this passed in tipset as well when expanding?

File: blockchain/chain/src/store/chain_store.rs (hunk 1 above)
Comment (austinabell) [reply blockchain/chain/src/store/chain_store.rs:107]: Well, I still don't see the reason why the mutex scope shouldn't be reduced, since the db never needs interior mutability. After reading the other parts a bit more I do see that the way it's setup forces a lock to be held for the whole function, when Cids are very cheap to clone. I just see the issue where each add, which happens after each block is validated, holds the lock while it performs io with the database during the function.

I don't think it's an issue with the current usage, but just thinking ahead to how this is setup. And yes we currently need this tracker to be used concurrently, as there will be multiple workers doing state transitions. That's why these types are setup to be threadsafe. We just have the default workers set to 1 for testing until everything is functional as there was some redundant overhead and made it harder to debug

**Review by austinabell — APPROVED** [review]
